### PR TITLE
✨ Allow changing DNSNameservers in subnet config for OpenstackCluster

### LINKED ISSUE #2468: Allow modifications to dnsNameservers in OpenstackCluster spec post-deployment
/kind feature

**Describe the solution you'd like**
Right now this is immutable:
```
spec:
  config
    managedSubnets:
    - cidr: 10.6.0.0/24
      dnsNameservers:
      - 1.1.1.1
      - 8.8.8.8
```
What if I would like to change the DNS servers after I have the cluster up?
In Openstack API it's a simple `openstack subnet` set call, but it's impossible through CAPO.

**Anything else you would like to add:**
Currently mutable fields:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/4142407d6bec0474e797850932b0584764b14e60/pkg/webhooks/openstackcluster_webhook.go#L76

### PR DESCRIPTION
**Pull request by pbasov**
**What this PR does / why we need it**:
Allows changing DNSNameservers in subnet config for OpenstackCluster after the initial deployment

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2468

/hold

### CODE REVIEW

**Comment by pbasov** [comment]
I absolutely hated making DNSNameservers mutable in the webhook, it was tricky because subnets is a list (even though we support only ONE per network). Surely there's a better way to manage mutability for the config.

**Comment by pbasov** [comment]
So yeah, why are we doing these webhook gymnastics when kubebuilder has immutable fields?
https://kubernetes.io/blog/2022/09/29/enforce-immutability-using-cel/#immutablility-after-first-modification

**Comment by lentzi90** [comment]
> So yeah, why are we doing these webhook gymnastics when kubebuilder has immutable fields? https://kubernetes.io/blog/2022/09/29/enforce-immutability-using-cel/#immutablility-after-first-modification

Likely because no-one has taken the time to refactor it yet

**Comment by pbasov** [comment]
@lentzi90 @mdbooth finished with the help of @s3rj1k, ready for review, tested manually on internal openstack cloud as well.

File: pkg/cloud/services/networking/network.go [hunk 1]
```diff
@@ -248,6 +253,39 @@ func (s *Service) createSubnet(openStackCluster *infrav1.OpenStackCluster, clust
...
+	// Picking the first managed subnet since we only support one for now
+	desiredNameservers := openStackCluster.Spec.ManagedSubnets[0].DNSNameservers
+	currentNameservers := subnet.DNSNameservers
+
+	var needsUpdate bool
+	if len(desiredNameservers) != len(currentNameservers) {
+		needsUpdate = true
+	} else {
+		needsUpdate = !equality.Semantic.DeepEqual(currentNameservers, desiredNameservers)
+	}
```
Comment (mnaser) [inline pkg/cloud/services/networking/network.go:267]: Probably not that big of a deal but I'm curious why don't we just drop `needsUpdate` and use the equality all the time.  Is this an optimization or is it simply something I don't know? :)

**Review by mnaser — APPROVED** [review]
looks good to me otherwise

**Review by lentzi90 — COMMENTED** [review]
/approve

File: pkg/webhooks/openstackcluster_webhook.go [hunk 2]
```diff
@@ -193,6 +194,62 @@ func (*openStackClusterWebhook) ValidateUpdate(_ context.Context, oldObjRaw, new
...
+
+				if newSubnet, exists := newSubnetMap[subnet.CIDR]; exists {
+					newSubnetCopy := newSubnet
+					newSubnetCopy.DNSNameservers = nil
+					newSubnets = append(newSubnets, newSubnetCopy)
+				}
+			}
+
+			oldObj.Spec.ManagedSubnets = oldSubnets
+			newObj.Spec.ManagedSubnets = newSubnets
```
Comment (mdbooth) [inline pkg/webhooks/openstackcluster_webhook.go:249]: We don't need this any more.

File: pkg/webhooks/openstackcluster_webhook.go [hunk 3]
```diff
@@ -193,6 +194,62 @@ func (*openStackClusterWebhook) ValidateUpdate(_ context.Context, oldObjRaw, new
...
+
+				// Check if AllocationPools have changed
+				if !equality.Semantic.DeepEqual(oldSubnet.AllocationPools, newSubnet.AllocationPools) {
+					allErrs = append(allErrs, field.Forbidden(
+						field.NewPath("spec", "managedSubnets").Child("allocationPools"),
+						"cannot modify allocation pools in existing subnet",
+					))
+				}
+
+				newSubnetMap[newSubnet.CIDR] = newSubnet
```
Comment (mdbooth) [inline pkg/webhooks/openstackcluster_webhook.go:230]:
```suggestion
				// DNSNameservers is mutable
				oldSubnet.DNSNameservers = nil
				newSubnet.DNSNameservers = nil
```

**Review by mdbooth — COMMENTED** [review]
Thanks for this, and especially for the amazing test coverage 👌

I'm not 100% confident of my assertion that order of DNSNameservers doesn't matter. It feels like one of those things that shouldn't matter, but might in practise. Let me know if I'm wrong about that.

Strong preference for an extra unit test covering an ordering-only difference, with the behaviour asserted being whatever we decide it should be.

Suggestion inline for a significant simplification of the webhook. Let me know if that makes sense.

File: pkg/cloud/services/networking/network.go (hunk 1 above)
Comment (s3rj1k) [reply pkg/cloud/services/networking/network.go:267]: `equality.Semantic.DeepEqual` is used in multiple places, so I guess it makes sense to have a wrapper function to compare slices, something like this

```
func EqualIgnoreOrder[T cmp.Ordered](a, b []T) bool {
	if len(a) != len(b) {
		return false
	}

	if len(a) == 0 {
		return true
	}

	aCopy := slices.Clone(a)
	bCopy := slices.Clone(b)

	slices.Sort(aCopy)
	slices.Sort(bCopy)

	return slices.Equal(aCopy, bCopy)
}
```

File: pkg/webhooks/openstackcluster_webhook.go (hunk 3 above)
Comment (s3rj1k) [reply pkg/webhooks/openstackcluster_webhook.go:230]: suggest to change to something like this, what do you think?

```
	// Allow changes only to DNSNameservers in ManagedSubnets spec
	if newObj.Spec.ManagedSubnets != nil && oldObj.Spec.ManagedSubnets != nil {
		// Check subnet count hasn't changed
		if len(oldObj.Spec.ManagedSubnets) != len(newObj.Spec.ManagedSubnets) {
			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "managedSubnets"), "cannot add or remove subnets"))

			return aggregateObjErrors(newObj.GroupVersionKind().GroupKind(), newObj.Name, allErrs)
		}

		// Build map of old subnets by CIDR
		oldSubnetMap := make(map[string]infrav1.SubnetSpec, len(oldObj.Spec.ManagedSubnets))
		for _, subnet := range oldObj.Spec.ManagedSubnets {
			oldSubnetMap[subnet.CIDR] = subnet
		}

		// Validate each new subnet
		for i, newSubnet := range newObj.Spec.ManagedSubnets {
			oldSubnet, exists := oldSubnetMap[newSubnet.CIDR]
			if !exists {
				allErrs = append(allErrs, field.Forbidden(
					field.NewPath("spec", "managedSubnets").Index(i).Child("cidr"),
					fmt.Sprintf("cannot change subnet CIDR to %s", newSubnet.CIDR),
				))

				continue
			}

			// Check AllocationPools haven't changed
			if !equality.Semantic.DeepEqual(oldSubnet.AllocationPools, newSubnet.AllocationPools) {
				allErrs = append(allErrs, field.Forbidden(
					field.NewPath("spec", "managedSubnets").Index(i).Child("allocationPools"),
					"cannot modify allocation pools in existing subnet",
				))
			}
		}

		// If there are errors, return early without modifying objects
		if len(allErrs) > 0 {
			return aggregateObjErrors(newObj.GroupVersionKind().GroupKind(), newObj.Name, allErrs)
		}

		// Clear DNSNameservers from both objects
		for i := range oldObj.Spec.ManagedSubnets {
			oldObj.Spec.ManagedSubnets[i].DNSNameservers = nil
		}

		for i := range newObj.Spec.ManagedSubnets {
			newObj.Spec.ManagedSubnets[i].DNSNameservers = nil
		}
	}
```

File: pkg/cloud/services/networking/network.go (hunk 1 above)
Comment (mdbooth) [reply pkg/cloud/services/networking/network.go:267]: This is just a nit, and mostly because I've spent time in the past reading Semantic.DeepEqual trying to determine the precise behaviour of its edge cases.

I only mention it at all because my personal preference would be to use something more explicit, especially when the data structure is simple as here. But this works and the PR has great test coverage, so there's no pressing need to update this.

Comment (s3rj1k) [reply pkg/cloud/services/networking/network.go:267]: I am fine replacing if we bring similar helper in, I would hate to replace 1 func call with len,copy,sort,equal checks in multiple places, this can be followup if that makes sense

File: pkg/webhooks/openstackcluster_webhook.go (hunk 3 above)
Comment (mdbooth) [reply pkg/webhooks/openstackcluster_webhook.go:230]: I was trying to get rid of the AllocationPools check. I think this is fragile because if we added another field to this struct we'd need to also check it hasn't changed. The chances of that being forgotten are high.

File: pkg/cloud/services/networking/network_test.go [hunk 4]
```diff
@@ -39,6 +39,86 @@ const (
...
+			desiredNameservers: []string{"8.8.8.8", "8.8.4.4"},
+			expect: func(m *mock.MockNetworkClientMockRecorder) {
+				m.UpdateSubnet(subnetID, subnets.UpdateOpts{
+					DNSNameservers: &[]string{"8.8.8.8", "8.8.4.4"},
+				}).Return(&subnets.Subnet{
+					ID:             subnetID,
+					DNSNameservers: []string{"8.8.8.8", "8.8.4.4"},
+				}, nil)
+			},
+		},
```
[the remark this answers is not in the document]
Comment (mdbooth) [reply pkg/cloud/services/networking/network_test.go:85]: We think that order probably does matter, so please can we have another test where the set of nameservers is identical but the order is different? We should assert that they're updated to be in the specified order.

**Comment by mdbooth** [comment]
This looks great. Can you squash the commits and I'll approve.

**Comment by s3rj1k** [comment]
> This looks great. Can you squash the commits and I'll approve.

Done, thanks for helping out and reviewing

**Review by lentzi90 — COMMENTED** [review]
/lgtm
/unhold
